### ISSUE TITLE
Custom function in expression builder displays parser error in layouts

### ISSUE THREAD

**Issue by fiddlersfan** [qgis/QGIS#42619]
**Describe the bug**
![Screenshot](https://user-images.githubusercontent.com/18557959/113334289-d44ebb00-9323-11eb-9960-aa97b81557bd.jpg)

**How to Reproduce**

1. Create new project
2. Create new layout
3. Add a label
4. Add a data defined override for e.g. elements width
5. Use a custom function in the expression string builder (e.g. the example custom function my_sum)
6. After the correct input the following error is still displayed
-> Parser Errors: syntax error, unexpected $end, expecting COMMA or ')'
Sometimes the "ok" button is greyed out ... this is not always reproducible

**QGIS and OS versions**

QGIS version | 3.16.5-Hannover | QGIS code revision | 58ba7c1ed6
-- | -- | -- | --
Compiled against Qt | 5.11.2 | Running against Qt | 5.11.2
Compiled against GDAL/OGR | 3.1.4 | Running against GDAL/OGR | 3.1.4
Compiled against GEOS | 3.8.1-CAPI-1.13.3 | Running against GEOS | 3.8.1-CAPI-1.13.3
Compiled against SQLite | 3.29.0 | Running against SQLite | 3.29.0
PostgreSQL Client Version | 11.5 | SpatiaLite Version | 4.3.0
QWT Version | 6.1.3 | QScintilla2 Version | 2.10.8
Compiled against PROJ | 6.3.2 | Running against PROJ | Rel. 6.3.2, May 1st, 2020
OS Version | Windows 10 (10.0)
Active python plugins | db_manager; MetaSearch; processing

**Additional context**
The error has only occurred since we started using the new LTR version (3.16) in the company. With the previous one (3.10) we had no problem with this. Is it possible that the error occurs in connection with the new preview feature in the expression string builder? Does this feature have a meaningful use in the layout context?

**Comment by roya0045** [comment]
what is your function?

**Comment by fiddlersfan** [comment]
I have used the example function that QGIS provides by default:
```
from qgis.core import *
from qgis.gui import *

@qgsfunction(args='auto', group='Custom')
def my_sum(value1, value2, feature, parent):
    """
    Calculates the sum of the two parameters value1 and value2.
    <h2>Example usage:</h2>
    <ul>
      <li>my_sum(5, 8) -> 13</li>
      <li>my_sum("field1", "field2") -> 42</li>
    </ul>
    """
    return value1 + value2
```

Important notice: If I use the same function outside the layout cotext the error is not displayed. An example of this is the data defined override for the symbol size of a vector layer.

![Screenshot 2021-04-02 103433](https://user-images.githubusercontent.com/18557959/113398553-123cf500-939f-11eb-9e70-320f16f5b3cf.png)

**Comment by roya0045** [comment]
...the test function requires a feature as input, in the dialogs you showed it has access to a feature but in the layout dialog it does not. This is clearly not a bug.

**Comment by fiddlersfan** [comment]
I can see the point but correct me if I'm wrong: a custom expression always gets the variables feature and parent passed to it. If you omit them in the code you get an argument parse error (...wrong number of arguments. Expected 0 but got 2.).

But this means in the end that no custom expressions can be used in the layout context or? 

I chose the function my_sum to keep the example simple. The custom expressions we use in the company are about the query of height and width of other layout elements by their id for a fluid layout.

**Comment by roya0045** [comment]
I don't have access to a version to try this, but what if you just use the *args, **kwargs way to catch anything in the declaration and just have the minimal number of arguments /keyword arguments otherwise? Does that work?

**Comment by fiddlersfan** [comment]
I can adjust the function declaration as suggested, but the error is still displayed....
For testing I have reinstalled the old LTR (3.10.14). The error is not displayed there for the indentical custom function. The output preview shows the right value without any errors.

![Screenshot 2021-04-02 163714](https://user-images.githubusercontent.com/18557959/113425414-1c2c1b80-93d2-11eb-9789-83e7ffa41687.jpg)

What is noticeable is that between both versions the feature selection and preview has been changed and expanded. My guess is that in this part the layout context is not fully caught and produced the error.

**Comment by roya0045** [comment]
partly, true, though the issue is that this check goes through https://github.com/qgis/QGIS/blob/2d1aa68f0d044f2aced7ebeca8d2fa6b754ac970/src/gui/qgsexpressionpreviewwidget.cpp#L85

Though something else is causing the parsing error of your function, and that is what may be triggering this error. Without the actual problematic function its hard to guess. If you can provide it would be helpful, or maybe a dev may be willing to take a deeper dive and ask you to provide it privately via email if confidentiality is a concern.

**Comment by fiddlersfan** [comment]
Fixed in #42745 for the default custom expression function _my_sum()_

As of version 3.16 it seems to be necessary to set the parameter _referenced_columns=[]_ in order to execute custom functions without reference to a feature without an error.
